action/search.go

```diff
@@ -13,52 +13,70 @@ import (
 
 // Search looks for packages with 'term' in their name.
 func Search(term, homedir string) {
-	charts, err := search(term, homedir)
+	charts, err := searchAll(term, homedir)
 	if err != nil {
 		log.Die(err.Error())
 	}
 
-	for dir, chart := range charts {
-		log.Info("\t%s (%s %s) - %s", filepath.Base(dir), chart.Name, chart.Version, chart.Description)
+	for _, name := range sortedIndex(charts) {
+		chart := charts[name]
+		log.Msg("\t%s (%s %s) - %s", name, chart.Name, chart.Version, chart.Description)
 	}
 }
 
-func search(term, homedir string) (map[string]*chart.Chartfile, error) {
-	files, err := filepath.Glob(filepath.Join(homedir, CacheChartPath, "*"))
+func sortedIndex(m map[string]*chart.Chartfile) []string {
+	ss := make(sort.StringSlice, len(m))
 
-	// only return chart directories
-	var dirs []string
-	for _, f := range files {
-		if filepath.Base(f) == ".git" {
-			continue
+	i := 0
+	for k, _ := range m {
+		ss[i] = k
+		i++
+	}
+
+	ss.Sort()
+	return ss
+}
+
+func searchAll(term, homedir string) (map[string]*chart.Chartfile, error) {
+	r := mustRepofile(homedir)
+	results := map[string]*chart.Chartfile{}
+	for _, table := range r.Tables {
+		tablename := table.Name
+		if table.Name == r.Default {
+			tablename = ""
 		}
-		fm, _ := os.Stat(f)
-		if !fm.IsDir() {
-			continue
+		base := filepath.Join(homedir, CachePath, table.Name, "*")
+		if err := search(term, base, tablename, results); err != nil {
+			log.Warn("Search error: %s", err)
 		}
-		dirs = append(dirs, f)
 	}
+	return results, nil
+}
 
+func search(term, base, table string, charts map[string]*chart.Chartfile) error {
+	dirs, err := filepath.Glob(base)
 	if err != nil {
-		return nil, fmt.Errorf("No results found. %s", err)
+		return fmt.Errorf("No results found. %s", err)
 	} else if len(dirs) == 0 {
-		return nil, errors.New("No results found.")
+		return errors.New("No results found.")
 	}
 
-	charts := make(map[string]*chart.Chartfile)
-
-	r, _ := regexp.Compile(term)
+	r, err := regexp.Compile(term)
+	if err != nil {
+		log.Die("Invalid expression %q: %s", term, err)
+	}
 
 	for _, dir := range dirs {
-		chart, err := chart.LoadChartfile(filepath.Join(dir, "Chart.yaml"))
+		cname := filepath.Join(table, filepath.Base(dir))
+		chrt, err := chart.LoadChartfile(filepath.Join(dir, "Chart.yaml"))
 
 		if err != nil {
-			log.Warn("failed to load Chart.yaml: %v", err)
+			//log.Warn("Failed to load Chart.yaml: %v", err)
```

Comment:
Right. It is not an error for a directory to not be a `chart`. This was generating a lot of spurious errors for directories like `.git`. (I actually removed the log message in one of the other PRs somewhere in here... so even the comment will go away)